app/uk/gov/hmrc/personaldetailsvalidation/endpoints/PersonalDetailsCollectionController.scala

```diff
@@ -16,19 +16,148 @@
+import uk.gov.hmrc.personaldetailsvalidation.model._
 import uk.gov.hmrc.personaldetailsvalidation.views.pages.PersonalDetailsPage
 import uk.gov.hmrc.play.bootstrap.controller.FrontendController
+import uk.gov.hmrc.personaldetailsvalidation.views.html.template._
+import uk.gov.hmrc.views.ViewConfig
+import java.time.LocalDate
+import cats.data.EitherT
+import scala.concurrent.Future
+import scala.util.Try
 
 class PersonalDetailsCollectionController @Inject()(page: PersonalDetailsPage,
-                                                    personalDetailsSubmission: FuturedPersonalDetailsSubmission)
-  extends FrontendController {
+                                                    personalDetailsSubmission: FuturedPersonalDetailsSubmission,
+                                                    appConfig: AppConfig)(implicit val messagesApi: MessagesApi, viewConfig: ViewConfig)
+  extends FrontendController with I18nSupport {
+
+  import uk.gov.hmrc.formmappings.Mappings._
+
+  private val initialForm: Form[InitialPersonalDetails] = Form(mapping(
+    "firstName" -> mandatoryText("personal-details.firstname.required"),
+    "lastName" -> mandatoryText("personal-details.lastname.required"),
+    "dateOfBirth" -> mandatoryLocalDate("personal-details")
+  )(InitialPersonalDetails.apply)(InitialPersonalDetails.unapply))
+
+  private def ninoValidation(): Mapping[Nino] = {
+    mandatoryText("personal-details.nino.required")
+      .verifying("personal-details.nino.invalid", nonEmptyString => Try(Nino(nonEmptyString.value.toUpperCase)).isSuccess)
+      .transform[Nino](validatedNonEmptyNino => Nino(validatedNonEmptyNino.value.toUpperCase), nino => NonEmptyString(nino.toString.toUpperCase))
+  }
+
+  private val pdvSessionKeys : List[String] = List("firstName", "lastName", "dob")
+
+  private lazy val ninoForm: Form[NinoDetails] = Form(mapping(
+    "nino" -> ninoValidation()
+  )(NinoDetails.apply)(NinoDetails.unapply))
+
+  private lazy val postcodeForm: Form[PostcodeDetails] = Form(mapping(
+    "postcode" -> postcodeValidation()
+  )(PostcodeDetails.apply)(PostcodeDetails.unapply))
+
+  private def postcodeValidation(): Mapping[NonEmptyString] = {
+    mandatoryText("personal-details.postcode.invalid").
+      verifying("personal-details.postcode.invalid", postcodeFormatValidation _)
+  }
+
+  private def postcodeFormatValidation(postcode: NonEmptyString) =
+    postcode.value.matches("""([A-Za-z][A-HJ-Ya-hj-y]?[0-9][A-Za-z0-9]?|[A-Za-z][A-HJ-Ya-hj-y][A-Za-z])\s?[0-9][ABDEFGHJLNPQRSTUWXYZabdefghjlnpqrstuwxyz]{2}""")
 
   def showPage(implicit completionUrl: CompletionUrl, alternativeVersion: Boolean): Action[AnyContent] = Action { implicit request =>
-    Ok(page.render(alternativeVersion))
+    if (appConfig.isMultiPageEnabled) {
+      val form: Form[InitialPersonalDetails] = (request.session.get("firstName"), request.session.get("lastName"), request.session.get("dob")) match {
+        case (Some(fn), Some(ln), Some(dob)) =>
```

Comment:
This grabbing of firstName lastName and dob out of the session is done lots of times. Stick it in a method perhaps?
```Request => (Option[String], Option[String], Option[String]) ```
The strings do make it a bit vulnerable to typos...